File: meridian-web/frontend/src/api/multiChannelBatchStt.ts
import api from './client';
import type { PublicSpeakerSide } from '../types';
import type { MultiChannelExportRequest } from './multiChannelExport';

// Этап 9.5: batch multi-channel STT (диагностический кандидат, live transcript не меняется).

export type MultiChannelBatchJobStatus =
  | 'queued' | 'preparing' | 'transcribing' | 'parsing' | 'comparing'
  | 'succeeded' | 'failed' | 'cancelled' | 'expired';

export interface MultiChannelBatchWord {
  text: string;
  start: number;
  end: number;
  channel_index: number;
  confidence: number | null;
  punctuated_word: string | null;
}

export interface MultiChannelBatchSegment {
  segment_id: string;
  channel_index: number;
  track_id: string;
  channel_label: string;
  side: string | null;
  text: string;
  start: number;
  end: number;
  confidence: number | null;
  words: MultiChannelBatchWord[];
}

export interface MultiChannelBatchChannel {
  channel_index: number;
  track_id: string;
  channel_label: string;
  side: string | null;
  source_kind: string;
  generation: number;
  transcript: string;
  words_count: number;
  segments_count: number;
  average_confidence: number | null;
  segments: MultiChannelBatchSegment[];
  warnings: string[];
}

export interface MultiChannelBatchResult {
  provider: string;
  model: string;
  language: string;
  provider_request_id: string | null;
  sample_rate: number;
  channels_count: number;
  duration_ms: number;
  channels: MultiChannelBatchChannel[];
  chronological_segments: MultiChannelBatchSegment[];
  combined_text: string;
  warnings: string[];
  provider_meta: Record<string, unknown>;
}

export interface MultiChannelBatchComparison {
  available: boolean;
  live_words: number;
  batch_words: number;
  live_chars: number;
  batch_chars: number;
  word_error_rate: number | null;
  text_similarity: number | null;
  channels_with_text: number;
  empty_channels: number[];
  average_confidence: number | null;
  overlap_segments: number;
  overlap_duration_ms: number;
  warnings: string[];
}

export interface MultiChannelBatchJob {
  job_id: string;
  meeting_id: number;
  status: MultiChannelBatchJobStatus;
  stage: string;
  progress: number;
  provider: string;
  model: string;
  language: string;
  created_at: string | null;
  started_at: string | null;
  finished_at: string | null;
  expires_at: string | null;
  result: MultiChannelBatchResult | null;
  comparison: MultiChannelBatchComparison | null;
  export_manifest: Record<string, unknown> | null;
  error_code: string | null;
  error_message: string | null;
  retryable: boolean;
}

export interface MultiChannelBatchSttRequest {
  export: MultiChannelExportRequest;
  channel_side_overrides?: Record<string, PublicSpeakerSide | null>;
  compare_with_live?: boolean;
}

export async function startMultiChannelBatchStt(
  meetingId: number, body: MultiChannelBatchSttRequest,
): Promise<MultiChannelBatchJob> {
  const { data } = await api.post<MultiChannelBatchJob>(
    `/meetings/${meetingId}/multi-source/batch-stt`, body);
  return data;
}

export async function getMultiChannelBatchSttJob(
  meetingId: number, jobId: string,
): Promise<MultiChannelBatchJob> {
  const { data } = await api.get<MultiChannelBatchJob>(
    `/meetings/${meetingId}/multi-source/batch-stt/${jobId}`);
  return data;
}

export async function cancelMultiChannelBatchSttJob(
  meetingId: number, jobId: string,
): Promise<void> {
  await api.delete(`/meetings/${meetingId}/multi-source/batch-stt/${jobId}`);
}

File: meridian-web/backend/app/schemas/multi_channel_batch_stt.py
"""Схемы batch multi-channel STT (Этап 9.5)."""

from pydantic import BaseModel, Field, model_validator

from .multi_channel_export import MultiChannelExportRequest


class MultiChannelBatchSttRequest(BaseModel):
    export: MultiChannelExportRequest
    channel_side_overrides: dict[str, str | None] = Field(default_factory=dict)
    compare_with_live: bool = True

    @model_validator(mode="after")
    def _validate(self):
        tids = self.export.track_ids
        # batch требует явный выбор >= 2 каналов (export-схема уже проверила unique/non-empty)
        if not tids or len(tids) < 2:
            raise ValueError("batch STT requires at least 2 selected tracks")
        selected = set(tids)
        bad = [k for k in self.channel_side_overrides if k not in selected]
        if bad:
            raise ValueError(f"channel_side_overrides reference unselected tracks: {bad}")
        return self


# --- typed outputs (для документации; сериализация в API через явные dict-билдеры) ---

class MultiChannelBatchWordOut(BaseModel):
    text: str
    start: float
    end: float
    channel_index: int
    confidence: float | None = None
    punctuated_word: str | None = None


class MultiChannelBatchSegmentOut(BaseModel):
    segment_id: str
    channel_index: int
    track_id: str
    channel_label: str
    side: str | None = None
    text: str
    start: float
    end: float
    confidence: float | None = None
    words: list[MultiChannelBatchWordOut] = Field(default_factory=list)


class MultiChannelBatchChannelOut(BaseModel):
    channel_index: int
    track_id: str
    channel_label: str
    side: str | None = None
    source_kind: str
    generation: int
    transcript: str
    words_count: int
    segments_count: int
    average_confidence: float | None = None
    segments: list[MultiChannelBatchSegmentOut] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


class MultiChannelBatchResultOut(BaseModel):
    provider: str
    model: str
    language: str
    provider_request_id: str | None = None
    sample_rate: int
    channels_count: int
    duration_ms: int
    channels: list[MultiChannelBatchChannelOut] = Field(default_factory=list)
    chronological_segments: list[MultiChannelBatchSegmentOut] = Field(default_factory=list)
    combined_text: str
    warnings: list[str] = Field(default_factory=list)
    provider_meta: dict = Field(default_factory=dict)


class MultiChannelBatchJobOut(BaseModel):
    job_id: str
    meeting_id: int
    status: str
    stage: str
    progress: float
    provider: str
    model: str
    language: str
    created_at: str | None = None
    started_at: str | None = None
    finished_at: str | None = None
    expires_at: str | None = None
    result: MultiChannelBatchResultOut | None = None
    comparison: dict | None = None
    export_manifest: dict | None = None
    error_code: str | None = None
    error_message: str | None = None
    retryable: bool = False
